### DISCUSSION TITLE
Creating Document Sets in a library

### DISCUSSION THREAD

**Discussion by rlv-dan** [pnp/pnpcore#1419] [Q&A] [answered]
I have create a content type based on the Document Set CT and added this to a library. How can I create a new document set in the library  and set the shared columns?

**Comment by jansenbe** [comment]
@rlv-dan : in the end a documentset is a folder in the SharePoint library, did you try uploading your files with their metadata in that folder? See https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#adding-files-uploading and https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#updating-file-metadata for generic guidance on that. 

Our docs is currently not explicitly showing this scenario, so please report back your findings so that we can update the docs

**Comment by rlv-dan** [comment, marked as answer]
Ok, here is how I solved it:

```
// Create a new folder
var list = await context.Web.Lists.GetByServerRelativeUrlAsync($"{context.Uri.PathAndQuery}/MyLibrary");
var folder = await list.RootFolder.EnsureFolderAsync("My Doc Set");

// Fetch the *list item* of the folder
await folder.ListItemAllFields.LoadAsync();
var listItem = await list.Items.GetByIdAsync(folder.ListItemAllFields.Id);

// Change the content type to one based on Document Set CT (= starts with 0x0120D520)
listItem["ContentTypeId"] = "0x0120D520005E02181D34C74530B814E626B873E036";
// Set shared fields
listItem["MySharedField"] = "123";
await listItem.UpdateAsync();
```
